From c32cdd17e0907b5f8288e1562c507fba94f507f2 Mon Sep 17 00:00:00 2001
From: Steve Loughran <stevel@apache.org>
Date: Tue, 21 Mar 2017 13:14:43 +0000
Subject: [PATCH 2411/2848] HADOOP-14204 S3A multipart commit failing,
 "UnsupportedOperationException at
 java.util.Collections$UnmodifiableList.sort".

Contributed by Steve Loughran

(cherry picked from commit 595501c9ddacd5fdb740ac1bf9bfeb0ffbbd6c0e)
(cherry picked from commit b4481fef41a815fb3e0d94a65fae48983a7e09b0)

Change-Id: Ibc61f844270833e3d58def160e5c24fe622441aa
---
 .../org/apache/hadoop/fs/s3a/S3AFileSystem.java    |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
index 56a7f4a..0cfaadd 100644
--- a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
+++ b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
@@ -2621,11 +2621,13 @@ CompleteMultipartUploadResult completeMultipartUpload(String uploadId,
           "No partitions have been uploaded");
       LOG.debug("Completing multipart upload {} with {} parts",
           uploadId, partETags.size());
+      // a copy of the list is required, so that the AWS SDK doesn't
+      // attempt to sort an unmodifiable list.
       return s3.completeMultipartUpload(
           new CompleteMultipartUploadRequest(bucket,
               key,
               uploadId,
-              partETags));
+              new ArrayList<>(partETags)));
     }
 
     /**
-- 
1.7.9.5

